concordium-consensus/src/Concordium/GlobalState/Persistent/Accounts.hs

```diff
@@ -168,35 +285,71 @@ getAccountByCredId cid accs@Accounts{..} =
         Nothing -> return Nothing
         Just ai -> fmap (ai,) <$> indexedAccount ai accs
 
--- | Get the account at a given index (if any).
-getAccountIndex :: (IsProtocolVersion pv, MonadBlobStore m) => AccountAddress -> Accounts pv -> m (Maybe AccountIndex)
-getAccountIndex addr Accounts{..} = AccountMap.lookup addr accountMap
+-- | Get the 'AccountIndex' for the provided 'AccountAddress' (if any).
+--  First try lookup in the in-memory difference map associated with the the provided 'Accounts pv',
+--  if no account could be looked up, then we fall back to the lmdb backed account map.
+--
+-- If account alises are supported then the equivalence class 'AccountAddressEq' is used for determining
+-- whether the provided @AccountAddress@ is in the map, otherwise we check for exactness.
+getAccountIndex :: forall pv m. (SupportsPersistentAccount pv m) => AccountAddress -> Accounts pv -> m (Maybe AccountIndex)
+getAccountIndex addr Accounts{..} = do
+    mAccountDiffMap <- liftIO $ readIORef accountDiffMapRef
+    case mAccountDiffMap of
+        Absent -> lookupDisk 0
+        Present accDiffMap ->
+            if supportsAccountAliases (protocolVersion @pv)
+                then
+                    DiffMap.lookupViaEquivalenceClass (accountAddressEmbed addr) accDiffMap >>= \case
+                        Just accIdx -> return $ Just accIdx
+                        Nothing -> do
+                            diffMapSize <- length <$> DiffMap.flatten accDiffMap
+                            lookupDisk $ fromIntegral diffMapSize
+                else
+                    DiffMap.lookupExact addr accDiffMap >>= \case
+                        Just accIdx -> return $ Just accIdx
+                        Nothing -> do
+                            diffMapSize <- length <$> DiffMap.flatten accDiffMap
+                            lookupDisk $ fromIntegral diffMapSize
+  where
+    -- Lookup the 'AccountIndex' in the lmdb backed account map,
+    -- and make sure it's within the bounds of the account table.
+    -- We do the bounds check as it could be that the lmdb backed account map
+    -- yields accounts which are not yet present in the @accountTable@.
+    -- In particular this can be the case if finalized blocks have been rolled
+    -- back as part of database recovery.
+    --
+    -- The extra @diffMapSize@ constraint is necessary in the case where a parent block to the last finalized block
+    -- is used for extending. This is normally not the case as blocks are typically built from the "best" block always.
+    --
+    -- However for scenarios like dry runs, then we need this extra constraint, as otherwise an account created in the
+    -- in the dry run would point to an account created in a successor of the block that the dry run is extending.
+    lookupDisk diffMapSize =
```

Comment:
It would be simpler to argue for the need for this diffMapSize on the basis of hygiene rather than this case by case basis.

It makes sense that if the account is not in the diffmap then it must have an index < #num of account - size of diffMap.